clio-client/src/main/scala/org/broadinstitute/clio/client/ClioClient.scala

```diff
@@ -1,61 +1,259 @@
+    */
+  sealed trait EarlyReturn
+  final case class ParsingError(error: String) extends EarlyReturn
+  final case class UsageOrHelpAsked(message: String) extends EarlyReturn
 
-  override val appVersion: String = ClioClientConfig.Version.value
+  val progName = "clio-client"
 
-  //not sure this is the best way to pass common opts but case-app docs are spotty at best
-  var commonOptions: CommonOptions = _
+  def main(args: Array[String]): Unit = {
+    implicit val system: ActorSystem = ActorSystem(progName)
+    import system.dispatcher
+    sys.addShutdownHook({ val _ = system.terminate() })
 
-  override def beforeCommand(options: CommonOptions,
-                             remainingArgs: Seq[String]): Unit = {
-    commonOptions = options
-    checkRemainingArgs(remainingArgs)
+    val webClient: ClioWebClient = new ClioWebClient(
+      ClioClientConfig.ClioServer.clioServerHostName,
+      ClioClientConfig.ClioServer.clioServerPort,
+      ClioClientConfig.ClioServer.clioServerUseHttps
+    )
+    val client: ClioClient = new ClioClient(webClient, IoUtil)
+
+    client
+      .instanceMain(args)
+      .fold(
+        {
+          case UsageOrHelpAsked(message) => {
+            println(message)
+            Future.successful(())
+          }
+          case ParsingError(error) => {
+            System.err.println(error)
+            Future.failed(new RuntimeException)
+          }
+        },
+        identity
+      )
+      .onComplete {
+        case Success(_) => sys.exit(0)
+        case Failure(_) => sys.exit(1)
+      }
   }
+}
 
-  private def checkRemainingArgs(remainingArgs: Seq[String]): Unit = {
-    if (remainingArgs.nonEmpty) {
-      logger.error(s"Found extra arguments: ${remainingArgs.mkString(" ")}")
-      sys.exit(1)
-    }
+/**
+  * The main clio-client application.
+  *
+  * Inspired by / partially copy-pasted from the
+  * [[caseapp.CommandAppWithPreCommand]], which demonstrates
+  * how to use caseapp's parsers to build a client with
+  * subcommands sharing common options. This class
+  * reimplements that class instead of extending it because
+  * that class is designed to be the top-level entry point of
+  * a program, with a main method and calls to sys.exit.
+  *
+  * Reimplementing the class also lets us tweak its logic so
+  * that common options are provided to the sub-command-handling
+  * method, rather than kept separate.
+  *
+  * @param webClient client handling HTTP communication with
+  *                  the clio-server
+  * @param ioUtil    utility handling all file operations,
+  *                  both local and in cloud storage
+  * @param ec        the execution context to use when scheduling
+  *                  asynchronous events
+  */
+class ClioClient(webClient: ClioWebClient,
+                 ioUtil: IoUtil)(implicit ec: ExecutionContext)
+    extends LazyLogging {
+  import ClioClient.{EarlyReturn, ParsingError, UsageOrHelpAsked}
+
+  val progName: String = ClioClient.progName
```

Comment:
not sure we need to store vals here if we are just passing them to copy and never using them again.